gitrevise/merge.py

```diff
@@ -199,29 +202,131 @@ def merge_blobs(
+
+
+def reuse_recorded_resolution(repo, preimage) -> Tuple[Optional[Blob], Optional[str]]:
+    rr_cache = repo.gitdir / "rr-cache"
+    # TODO Should we cache settings like this?
+    if not repo.bool_config("rerere.enabled", default=rr_cache.is_dir()):
+        return None, None
+
+    conflict_id = conflict_id_by_file_contents(preimage)
+    conflict_dir = rr_cache / conflict_id
+    if not conflict_dir.is_dir():
+        return None, conflict_id
+    if not repo.bool_config("rerere.autoUpdate", default=False):
+        return None, conflict_id
+    postimage_path = conflict_dir / "postimage"
+    postimage = postimage_path.read_bytes()
+    recorded_preimage = (conflict_dir / "preimage").read_bytes()
+    try:
+        result = merge_blobs(
+            repo.workdir,  # TODO is this correct?
+            labels=(
+                "current",
+                "base",
+                "other",
+            ),
+            current=Blob(repo, postimage),
+            base=Blob(repo, recorded_preimage),
+            other=Blob(repo, preimage),
+        )
+    except MergeConflict:
+        # We could ask the user to merge this. However, that could be confusing.
+        # Just fall back to letting them resolve the entire conflict.
+        return None, None
+    print("Successfully replayed recorded resolution")
+    # Mark that "postimage" was used to help git gc (git does the same).
+    os.utime(postimage_path)
+    return result, conflict_id
+
+
+def record_resolution(
+    repo: Repository, conflict_id: Optional[str], preimage: bytes, postimage: bytes
+) -> None:
+    if conflict_id is None:
+        return
+
+    conflict_dir = repo.gitdir / "rr-cache" / conflict_id
+    conflict_dir.parent.mkdir(exist_ok=True)
+    conflict_dir.mkdir(exist_ok=True)
+    print("Recording conflict resolution")
+    (conflict_dir / "preimage").write_bytes(preimage)
+    (conflict_dir / "postimage").write_bytes(postimage)
+
+
+RR_OTHER = 0
+RR_SIDE_1 = 1
+RR_SIDE_2 = 2
+
+
+def conflict_id_by_file_contents(preimage: bytes) -> str:
+    hunk = RR_OTHER
+    one = b""
+    two = b""
+    context = hashlib.sha1()
+    for line in preimage.splitlines():
+        if line.startswith(b"<" * MARKER_SIZE):
+            hunk = RR_SIDE_1
+            continue
+        if line.startswith(b"|" * MARKER_SIZE):
+            if hunk != RR_SIDE_1:
+                break
+            hunk = RR_OTHER
+            continue
+        if line.startswith(b"=" * MARKER_SIZE):
+            hunk = RR_SIDE_2
+            continue
+        if line.startswith(b">" * MARKER_SIZE):
+            if one > two:
+                one, two = two, one
+            context.update(one + b"\0" + two + b"\0")
+            one = b""
+            two = b""
+        if hunk == RR_SIDE_1:
+            one += line + b"\n"
+            continue
+        if hunk == RR_OTHER:
+            continue
+        if hunk == RR_SIDE_2:
+            two += line + b"\n"
+            continue
+        assert False
```

Comment:
This doesn't quite match the behaviour of git's hashing logic from `handle_path`. It doesn't handle edge cases like nested conflict markers exactly the same as the one used by git itself.

I think the easiest way to implement something like that would be to mimic the original structure as closely as possible. This would also provide the normalization logic required for writing out the recorded preimage, and performing merges. It'd probably look something like (again, largely untested):

```python
class ConflictParseFailed(Exception):
    pass


def normalize_conflict(
        lines: Iterator[bytes],
        marker_size: int,
        hasher: Optional["hashlib._Hash"],
) -> bytes:
    cur_hunk: Optional[bytes] = b""
    other_hunk: Optional[bytes] = None
    while True:
        line = next(lines, None)
        if line is None:
            raise ConflictParseFailed("unexpected eof")
        elif line.startswith(b"<" * marker_size):
            # parse recursive conflicts, including their processed output in the current hunk
            conflict = normalize_conflict(lines, marker_size, None)
            if cur_hunk is not None:
                cur_hunk += conflict
        elif line.startswith(b"|" * marker_size):
            # ignore the diff3 ||||||| section. Must be still parsing the first hunk.
            if other_hunk is not None:
                raise ConflictParseFailed("unexpected ||||||| conflict marker")
            (other_hunk, cur_hunk) = (cur_hunk, None)
        elif line.startswith(b"=" * marker_size):
            # switch into the second hunk after the ======= marker
            # could be in either the diff3 original section or the first hunk
            if cur_hunk is not None:
                if other_hunk is not None:
                    raise ConflictParseFailed(
                        "unexpected ======= conflict marker")
                other_hunk = cur_hunk
            cur_hunk = b""
        elif line.startswith(b">" * marker_size):
            # end of conflict. update hasher, and return a normalized conflict
            if cur_hunk is None or other_hunk is None:
                raise ConflictParseFailed("unexpected >>>>>>> conflict marker")

            hunk1, hunk2 = sorted([cur_hunk, other_hunk])
            if hasher:
                hasher.update(hunk1 + b"\0")
                hasher.update(hunk2 + b"\0")
            return b"".join((
                b"<" * marker_size + b"\n",
                hunk1,
                b"=" * marker_size + b"\n",
                hunk2,
                b">" * marker_size + b"\n",
            ))
        elif cur_hunk is not None:
            # add non-marker lines to the current hunk (or discard if in the diff3 hunk)
            cur_hunk += line


def normalize_conflicted_file(body: bytes,
                              marker_size: int = 7) -> Tuple[bytes, str]:
    hasher = hashlib.sha1()
    normalized = b""

    lines = iter(body.splitlines(keepends=True))
    while True:
        line = next(lines, None)
        print(f"line = {line}")
        if line is None:
            return normalized, hasher.hexdigest()
        elif line.startswith(b"<" * marker_size):
            normalized += normalize_conflict(lines, marker_size, hasher)
        else:
            normalized += line
```